File: combination-sum.py
# -*- coding: utf-8 -*-

"""
https://leetcode.com/problems/combination-sum/


Given a set of candidate numbers (candidates) (without duplicates) and a target number (target),
find all unique combinations in candidates where the candidate numbers sums to target.

The same repeated number may be chosen from candidates unlimited number of times.

Note:
All numbers (including target) will be positive integers.
The solution set must not contain duplicate combinations.

Example 1:
Input: candidates = [2,3,6,7], target = 7,
A solution set is:
[
  [7],
  [2,2,3]
]

Example 2:
Input: candidates = [2,3,5], target = 8,
A solution set is:
[
  [2,2,2,2],
  [2,3,3],
  [3,5]
]
"""


class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        # Sort candidates to avoid duplicate recurse
        sort_candidates = sorted(candidates)
        result = []
        # Recurse
        self.recurse(sort_candidates, target, [], result)
        return result

    def recurse(self, candidates, target, tmp_nums, result):
        # If target is 0, then tmp_nums[] is a valid combination.
        if target == 0:
            result.append(tmp_nums[:])
            return
        # If candidates is empty, then return.
        if not candidates:
            return
        # Traverse in candidates
        for i in range(len(candidates)):
            # Jump the same candidates to avoid duplicate recurse
            if i > 0 and candidates[i] == candidates[i - 1]:
                continue
            if candidates[i] <= target:
                # The candidates[i + 1:] is next candidates
                self.recurse(candidates[i:], target - candidates[i], tmp_nums + [candidates[i]], result)
            else:
                break


if __name__ == '__main__':
    s = Solution()
    print s.combinationSum([2,3,6,7], 7)

File: combination-sum.cpp
/*
Given a set of candidate numbers (C) and a target number (T), find all unique combinations in C where the candidate numbers sums to T.
The same repeated number may be chosen from C unlimited number of times.

Note:
All numbers (including target) will be positive integers.
Elements in a combination (a1, a2, … , ak) must be in non-descending order. (ie, a1 ≤ a2 ≤ … ≤ ak).
The solution set must not contain duplicate combinations.
For example, given candidate set 2,3,6,7 and target 7, 
A solution set is: 
[7] 
[2, 2, 3] 

Analysis:
回溯法解决组合问题，由于考虑到重复数据的问题，需要对数据进行排序。
 */
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int> > combinationSum(vector<int>& candidates, int target) {
    	vector<vector<int> > res;
    	vector<int> tmp;
    	sort(candidates.begin(), candidates.end());
    	combinate(candidates, 0, target, tmp, res);
    	return res;
    }

    void combinate(vector<int>& c, int pos, int target, vector<int>& tmp, vector<vector<int> >& res) {
    	if (target == 0 && tmp.size() > 0) {
    		res.push_back(tmp);
    		return;
    	}

    	for (int i = pos; i < c.size(); ++i) {
    		if (target < c[i])
    			break;
    		if (i > 0 && c[i] == c[i] - 1)
    			continue;
    		tmp.push_back(c[i]);
    		combinate(c, i, target - c[i], tmp, res);
    		tmp.pop_back();
    	}
    }
};

int main(int argc, char* argv)
{
	Solution a;
	int s[] = {2, 3, 6, 7};
	vector<int> candidates (s, s + sizeof(s)/sizeof(int));
	vector<vector<int> > res = a.combinationSum(candidates, 7);
	for (int i = 0; i < res.size(); ++i) {
		for (int j = 0; j < res[i].size(); ++j) {
			cout<<res[i][j]<<" ";
		}
		cout<<endl;
	}
	return 0;
}


